Federating gathio

dariusk opened the pull request:
Here it is, my giant federation pull request! Please don't merge, I'm sure I will find more problems, but I would love for you to look at this, answer my questions below, and so on.

### Changes

* added `activitypub.js`, which contains all activitypub-related functions
* added `FEDERATION.md`, intended to be a reference for ActivityPub developers who want to make their service compatible with gathio, [see here for more](https://socialhub.activitypub.rocks/t/documenting-federation-behavior-in-a-semi-standard-way/453/7)
* created a `helpers.js` file and moved `addToLog()` there
* added a config to Express so it can serve JSON as needed
* added a Followers model, to hold followers of an event until the event is over
* added an ActivityPubMessages model, which holds all messages sent to activitypub servers by the server (until the event is deleted)
* added activityPub fields to EventSchema and CommentSchema
* added new routes to the API:
  * GET `/:eventID/featured` returns JSON for the featured/pinned post of an event
  * GET `/:eventID/m/:hash` returns the JSON of a previously-sent ActivityPub message with the associated hash
  * GET `/.well-known/webfinger` returns the JSON of a webfinger object required for the initial handshake with a remote ActivityPub server
* some standard gathio functions now trigger ActivityPub activities
  * for example, scheduled deletion sends a Delete notice to all subscribers so they can remove it from their databases too
  * basically anything that sends an email to users now also sends an ActivityPub message to any activitypub users
  * (there are others too)
* almost all instances of "Gathio" or "gath.io" as a site name are now replaced with `siteName` from the `config/domain.js` file
* almost all instances of "gath.io" as a domain are now replaced with `domain` from the `config/domain.js` file
* the Node.js version is now specified in `package.json`

There is now a toggle in the `config/domain-example.js` file called `isFederated`. When set to `true`, ActivityPub functions as normal. When set to false, the ActivityPub functions all immediately return without doing anything, and AP routes return 404s.

### A bug?

I noticed this code:

````
if (req.files && Object.keys(req.files).length != 0) {
		importediCalObject = ical.parseICS(req.files.icsImportControl.data.toString('utf8'));
		for (var key in importediCalObject) {
    		importedEventData = importediCalObject[key];
		}
		//...//
	}
````

It seems to overwrite `importedEventData` with every individual key, which is the opposite of what you wanted, I think? I changed it to

```
if (req.files && Object.keys(req.files).length !== 0) {
		let importediCalObject = ical.parseICS(req.files.icsImportControl.data.toString('utf8'));
    let importedEventData = importediCalObject;
    //...//
	}
```

Let me know if that's an error on my part and I'll remove it from the PR.

### Tabs and spaces

I noticed there wasn't consistent use of tabs vs spaces for indents. My code uses spaces generally but I can switch it all to tabs if you prefer.

My first instinct was to automatically convert everything but it made the diffs too difficult to read, so maybe after this PR we can do a tab/space cleanup on the entire codebase (there are ways to automate this, thankfully).

dariusk commented:
Also I'm going to be putting a walkthrough of how this all works on the wiki, which you'll probably want to see! Expect it tomorrow.

lowercasename commented:
This looks great - I'll probably not have time to look at it until next week, but also I'm not familiar with ActivityPub so if anyone with more specific knowledge wants to also have a look, please get them on board!

I'll check that bug over soon as I can x

I absolutely prefer spaces, I think this is a consequence of switching editors at one stage. We should definitely automatically switch it over.

dariusk commented:
Hey @wxcafe -- care to take a quick look at this? @lowercasename isn't familiar with ActivityPub so can't really comment on a lot of this, in particular the `activitypub.js` file.

lowercasename commented:
I've had a look at that bug and what the code appears to have been doing is taking an object with one nested object and removing the overlying object, so it wasn't a bug, just weird. I've made it more intelligible in this latest commit, which I'm afraid I made a bit of a dog's breakfast of: https://github.com/lowercasename/gathio/commit/7877e73537a0ce0650cc6bb073c9a81432bd35e1

So if you remove any changes to the iCal import code from your commit I think we should be right. I also couldn't help it and put some code in for a logo in the email, which I really should have been doing in your branch - again, sorry for making a mess of it! This is what happens when I don't pay attention to git.

dariusk commented:
I reverted the ical import stuff in fa2ca34a5c8f0e10a902129caec36dea38eec13a, it's now back to what you had. I've also merged the latest master branch and tested it and it should be good to go.

I wasn't able to find someone else to do code review. The main remaining task is to write up this wiki page:

https://github.com/lowercasename/gathio/wiki/Fediverse-Instructions

This is the page that is referenced in this screenshot, near the bottom, and acts as a kind of tutorial for Fediverse users who haven't encountered a gathio event in the wild.

<img width="597" alt="image" src="https://user-images.githubusercontent.com/266454/72230290-37aacc80-3569-11ea-9835-26ff20dbda09.png">

wxcafe commented:
Sorry, work was on fire last week. I’ll try to take a look ASAP

Clément 'wxcafé' Hertling

> On Jan 12, 2020, at 9:28 PM, Darius Kazemi <notifications@github.com> wrote:
> 
> ﻿
> I reverted the ical import stuff in fa2ca34, it's now back to what you had. I've also merged the latest master branch and tested it and it should be good to go.
> 
> I wasn't able to find someone else to do code review. The main remaining task is to write up this wiki page:
> 
> https://github.com/lowercasename/gathio/wiki/Fediverse-Instructions
> 
> This is the page that is referenced in this screenshot, near the bottom, and acts as a kind of tutorial for Fediverse users who haven't encountered a gathio event in the wild.
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

dariusk commented:
I have updated the wiki with instructions. They're not perfect but they're a start.

https://github.com/lowercasename/gathio/wiki/Fediverse-Instructions

Pending a look from @wxcafe (thank you!!!) I consider this ready to go. (After this merges I have another feature pull request, and I'll make one that converts tabs to spaces as well.)

dariusk commented on views/home.handlebars:
```diff
@@ -1,4 +1,4 @@
-<h1>Organise all the things</h1>
+<h1>{{siteName}}</h1>
```
This is one that I specifically changed the semantic meaning of to "site name", because it lets the "gath.io" branding coexist alongside whatever the name of your instance is.

dariusk commented:
So the traceback that @wxcafe got of

```
Jan 13 21:33:18 events npm[2576]: Error: ReferenceError: URL is not defined
Jan 13 21:33:18 events npm[2576]:     at _handleFollow (/srv/gathio/activitypub.js:464:17)
Jan 13 21:33:18 events npm[2576]:     at Object.processInbox (/srv/gathio/activitypub.js:855:7)
```

means that he is running a version of Node that doesn't have URL defined as a global. I'm adding a specification to the `package.json` that you need to be running Node 10 LTS or later. That should fix the follows issue.

Also in a future commit I will add a toggle in the `config/domain.js` file to turn off federation completely.

palfrey commented:
BTW, adding `RUN apk add --no-cache python3 build-base` to the Dockerfile below the `WORKDIR` line should make this build again.

dariusk commented:
Okay, I've added the toggle to the `config/domain-example.js`! I think this might be ready to go??

wxcafe commented:
Great! A few more points:
- there's a few cryptic debug messages (`Response: 202` / `sent to [...]`), these should probably be either expanded or removed
- we need updated docs for node10, but yeah it works perfectly there
- old events aren't federated once federation is turned on (no actors created for them) but the fediverse links are there. You can copy them but they don't do anything.
- should federation be default off? I'd say on is a better default, wdyt?

dariusk commented:
Oh right. I should account for a case where someone toggles federation after a server has been in production for a while. Will do....

I'll handle the docs and debug message stuff.

I think federation should be on as a default. The flagship instance at https://gath.io will be federated so I think it's reasonable to assume that defaults should be equivalent to what is there.

wxcafe commented:
Great, thanks!

Yeah so do I but it's not right now jsyk :)

wxcafe commented:
Great, thanks!

Yeah so do I but it's not right now jsyk :)

dariusk commented:
Wait really? `isFederated` is set to true in the example config file.

wxcafe commented:
Yeah I meant, if it’s not present in the config file it defaults to false

Clément 'wxcafé' Hertling

> On Jan 19, 2020, at 8:23 PM, Darius Kazemi <notifications@github.com> wrote:
> 
> ﻿
> Wait really? isFederated is set to true in the example config file.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

dariusk commented:
Okay, now when there is not an `isFederated` variable present in the config, federation is enabled. You must explicitly set `isFederated` to `false` in the config to turn it off.

I changed how the toggle works. When federation is turned off, the database records for federation are still created (actors and so on) but the routes and internal events that trigger federation messages being sent out are turned off. This way when it's toggled back on, the actors for events are immediately accessible from other AP servers.

Normally you wouldn't want to create database entries for a feature that isn't even being used, but since the whole point of gath is to delete the database extremely regularly, it shouldn't be an issue.

Also fixed up some of the cryptic debug messages. Updated the installation instructions to specify the Node.js version.

I believe this resolves all outstanding suggestions. (Again @wxcafe thank you so much for your review here, it's been really helpful.)

dariusk commented on routes.js:
```diff
@@ -14,63 +14,68 @@ const router = express.Router();
-        }
-        return String.fromCharCode(escapeCode.match(/[0-9]+/));
-    });
+function htmlEscapeToText (text) {
```
Created an issue for this https://github.com/lowercasename/gathio/issues/32

lowercasename commented:
This looks great, all, I'm delighted to merge it. Really makes Gathio feel like a more mature piece of software.

wxcafe commented:
:+1: :)

lowercasename commented:
Another question, should there be a different icon for the federation link? I suppose a Mastodon icon isn't general enough? Any good ideas appreciated.

dariusk commented:
I thought about this and decided to keep it generic. My thought process is that there are two possible icons that make sense. One is the ActivityPub icon:

<img width="244" alt="image" src="https://user-images.githubusercontent.com/266454/72821168-fc716300-3c24-11ea-8631-8fbd60cc59af.png">

In my mind this would be like putting the RSS icon on a page, so there is precedent for that.

The other option is this Fediverse icon, which is an unofficial one used by some of the community:

<img width="321" alt="image" src="https://user-images.githubusercontent.com/266454/72821272-24f95d00-3c25-11ea-9d3f-8863facda39d.png">

My problem with it is... it's bad

I wouldn't be opposed to using the AP icon but I think it's almost as good to just use the generic share icon.

lowercasename commented:
Yes that second symbol is... bad. I think it's fine to leave the share icon for now, although it might be a bit confusing for people who don't know what that string of symbols relates to. But an ActivityPub icon (I like the first one!) won't really make it any clearer.

dariusk commented:
I was considering a `?` you can click and get a little modal with some help text like, "share this with people to let them connect to you on the fediverse. [learn more]"
